### DISCUSSION TITLE
Pipeline design with dynamic partitions

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#18275] [Q&A]
Hi,
I wonder how to construct this :
• I parse few xml files in which I extract Document path and edited timestamp. The output is a pandas dataframe in which the number of rows is not known in advance
• Then I want to extract info from this Documents 
How would you construct this ?
• I first start to use StaticPartitionsDefinition(xml_list)
• Then create this assest : @asset(partitions_def=sitemaps_partitions_defs) this return the dataframe
• Then DynamicPartitionsDefinition for the second asset ?

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1700907170596899?thread_ts=1700907170.596899&cid=C01U954MEER).*

**Comment by smackesey** [comment]
IIUC you have:

- A set of XML documents with names known in advance
- Each XML document defines a set of documents (of unknown identity/number before reading the XML file)

I think it makes sense to use `StaticPartitionsDefinition` for your asset that converts an XML document to a pandas Dataframe. Thereafter, whether to use a partition to represent each document depends on how you'd like to process the documents-- if you want to process each document in a separate run, then that makes sense.

@clairelin135 is our resident partitions expert-- Claire is it a well-supported pattern to derive a set of dynamic partitions from the value of an upstream asset?

**Reply by clairelin135** [reply]
+1 on Sean's comments above, I agree it depends on the use case:

If you want to process each pandas dataframe in one run, I would recommend using [dynamic fan outs in a graph-backed asset](https://github.com/dagster-io/dagster/discussions/15521).

We do support deriving dynamic partitions from the output of an upstream asset. However, setting up a dependency relationship between an upstream static partitioned asset and downstream dynamic partitioned asset is not well supported, since you currently can't map a dynamic partition to a specific upstream static partition (https://github.com/dagster-io/dagster/issues/13139).

For that reason I would recommend processing each pandas dataframe in its own run via dynamic fan outs. Happy to brainstorm other workarounds if that approach doesn't work for a particular reason.

**Reply by parkerdaminer** [reply]
What about a multi_asset? Have a defined multi_asset that reads the XML and determines the different assetMaterializations and Outputs needed based on the XML document subsets.

@clairelin135 does specify that dynamic fan out example doesn't fulfill the partitioned asset portion of the question.
